torch/utils/checkpoint.py

```diff
@@ -0,0 +1,140 @@
+import torch
+from torch.autograd import Variable, Function
+
+
+def wrap_variable(inputs):
+    if torch.is_tensor(inputs):
+        return Variable(inputs)
+    elif isinstance(inputs, tuple):
+        return tuple(wrap_variable(v) for v in inputs)
+    else:
+        raise RuntimeError("Unsupported input type: ", type(inputs).__name__)
+
+
+def unpack_variables(inputs):
+    if isinstance(inputs, Variable):
+        return inputs.data
+    elif torch.is_tensor(inputs):
+        return inputs
+    elif isinstance(inputs, tuple):
+        return tuple(unpack_variables(v) for v in inputs)
+    else:
+        raise RuntimeError("Unsupported input type: ", type(inputs).__name__)
+
+
+class CheckpointFunction(Function):
+
+    @staticmethod
+    def forward(ctx, run_function, *args):
+        ctx.run_function = run_function
+        ctx.save_for_backward(*args)
+        var_args = wrap_variable(args)
+        with torch.no_grad():
+            outputs = run_function(*var_args)
+        return unpack_variables(outputs)
+
+    @staticmethod
+    def backward(ctx, *grads):
+        real_inputs = ctx.saved_variables
+        # We need to create new Variables to mark this place in the graph.
+        # Reusing real_inputs would be incorrect if a case like this:
+        #
+        # y = checkpoint(lambda x: x + 1, x)
+        # z = checkpoint(lambda x, y: x + y, x, y)
+        #
+        # This would fail, because when grad((x + y), (x, y)) is called in
+        # the second checkpoint, autograd would traverse all paths from (x + y)
+        # to the definition of x, which includes the first checkpoint. To
+        # prevent this situation, we create views of the inputs, which lets us
+        # still get all correctness checks, but uniquely marks the place up to
+        # which we want to differentiate, because all views are independent nodes
+        # (i.e. there is no path from one to another via .grad_fn chain).
+        inputs = [i[:] for i in real_inputs]
+        with torch.enable_grad():
+            outputs = ctx.run_function(*inputs)
+        if isinstance(outputs, Variable)
```

Comment:
There should be a `:` at the end of line

Reply:
Thanks, I've fixed that